File: app/src/main/java/com/shrikantbadwaik/tmdb/data/remote/apiresponse/ImageResponse.java
package com.shrikantbadwaik.tmdb.data.remote.apiresponse;

import android.os.Parcel;
import android.os.Parcelable;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.shrikantbadwaik.tmdb.data.model.Image;

import java.util.ArrayList;
import java.util.List;

public class ImageResponse implements Parcelable {
    public static final Creator<ImageResponse> CREATOR = new Creator<ImageResponse>() {
        @Override
        public ImageResponse createFromParcel(Parcel in) {
            return new ImageResponse(in);
        }

        @Override
        public ImageResponse[] newArray(int size) {
            return new ImageResponse[size];
        }
    };

    @JsonProperty("id")
    private String movieId;
    @JsonProperty("backdrops")
    private List<Image> backdropList;
    @JsonProperty("posters")
    private List<Image> posterList;

    public ImageResponse() {
        backdropList = new ArrayList<>();
        posterList = new ArrayList<>();
    }

    protected ImageResponse(Parcel in) {
        movieId = in.readString();
        backdropList = in.createTypedArrayList(Image.CREATOR);
        posterList = in.createTypedArrayList(Image.CREATOR);
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(movieId);
        dest.writeTypedList(backdropList);
        dest.writeTypedList(posterList);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public String getMovieId() {
        return movieId;
    }

    public void setMovieId(String movieId) {
        this.movieId = movieId;
    }

    public List<Image> getBackdropList() {
        return backdropList;
    }

    public void setBackdropList(List<Image> backdropList) {
        this.backdropList = backdropList;
    }

    public List<Image> getPosterList() {
        return posterList;
    }

    public void setPosterList(List<Image> posterList) {
        this.posterList = posterList;
    }
}

File: app/src/main/java/com/shrikantbadwaik/tmdb/data/remote/apiresponse/ImageResponse.kt
package com.shrikantbadwaik.tmdb.data.remote.apiresponse

import android.os.Parcel
import android.os.Parcelable
import com.fasterxml.jackson.annotation.JsonProperty
import com.shrikantbadwaik.tmdb.data.model.Image

data class ImageResponse(
    @JsonProperty("id")
    val movieId: String?,
    @JsonProperty("backdrops")
    val backdropList: MutableList<Image>?,
    @JsonProperty("posters")
    val posterList: MutableList<Image>?
) : Parcelable {
    constructor(parcel: Parcel) : this(
        parcel.readString(),
        parcel.createTypedArrayList(Image),
        parcel.createTypedArrayList(Image)
    )

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(movieId)
        parcel.writeTypedList(backdropList)
        parcel.writeTypedList(posterList)
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun toString(): String {
        return "ImageResponse(movieId=$movieId, backdropList=$backdropList, posterList=$posterList)"
    }

    companion object CREATOR : Parcelable.Creator<ImageResponse> {
        override fun createFromParcel(parcel: Parcel): ImageResponse {
            return ImageResponse(parcel)
        }

        override fun newArray(size: Int): Array<ImageResponse?> {
            return arrayOfNulls(size)
        }
    }


}